Libraries/WebSocket/WebSocket.js

```diff
@@ -0,0 +1,148 @@
+var MessageEvent = require('MessageEvent');
+
+var WebSocketId = 0;
+
+var CLOSE_NORMAL = 1000;
+
+class WebSocket extends WebSocketBase {
+  _socketId: number;
+  _subs: any;
+
+  connectToSocketImpl(url: string): void {
+    this._socketId = WebSocketId++;
+
+    RCTWebSocketModule.connect(url, this._socketId);
+
+    this._registerEvents(this._socketId);
+  }
+
+  closeConnectionImpl(code?: number, reason?: string): void {
+    this._closeWebSocket(this._socketId, code, reason);
+  }
+
+  cancelConnectionImpl(): void {
+    this._closeWebSocket(this._socketId);
+  }
+
+  sendStringImpl(message: string): void {
+    RCTWebSocketModule.send(message, this._socketId);
+  }
+
+  sendArrayBufferImpl(): void {
+    // TODO
+    console.warn('Sending ArrayBuffers is not yet supported');
+  }
+
+  _closeWebSocket(id: number, code?: number, reason?: string): void {
+    if (Platform.OS === 'android') {
+      /*
+       * The status code 1000 means 'CLOSE_NORMAL'
+       * Reason is empty string by to match browser behaviour
+       * More info: https://developer.mozilla.org/en-US/docs/Web/API/CloseEvent
+       */
+      var statusCode = typeof code === 'number' ? code : CLOSE_NORMAL;
+      var closeReason = typeof reason === 'string' ? reason : '';
+
+      RCTWebSocketModule.close(statusCode, closeReason, id);
+    } else {
+      RCTWebSocketModule.close(id);
+    }
+  }
+
+  _unregisterEvents(): void {
+    this._subs.forEach(e => e.remove());
+    this._subs = [];
+  }
+
+  _registerEvents(id: number): void {
+    this._subs = [
+      RCTDeviceEventEmitter.addListener('websocketMessage', ev => {
+        if (ev.id !== id) {
+          return;
+        }
+
+        var event = new MessageEvent('message', {
+          data: ev.data
+        });
+
+        this.onmessage && this.onmessage(event);
+        this.dispatchEvent(event);
+      }),
+      RCTDeviceEventEmitter.addListener('websocketOpen', ev => {
+        if (ev.id !== id) {
+          return;
+        }
+
+        this.readyState = this.OPEN;
+
+        var event = new Event('open');
+
+        this.onopen && this.onopen(event);
+        this.dispatchEvent(event);
+      }),
+      RCTDeviceEventEmitter.addListener('websocketClosed', ev => {
+        if (ev.id !== id) {
+          return;
+        }
+
+        this.readyState = this.CLOSED;
+
+        var event = new Event('close');
+
+        this.onclose && this.onclose(event);
+        this.dispatchEvent(event);
+        this._unregisterEvents();
+        this._closeWebSocket(id);
+      }),
+      RCTDeviceEventEmitter.addListener('websocketFailed', ev => {
+        if (ev.id !== id) {
+          return;
+        }
+
+        var event = new Event('error');
+
+        event.message = ev.message;
+
+        this.onerror && this.onerror(event);
+        this.dispatchEvent(event);
+        this._unregisterEvents();
+        this._closeWebSocket(id);
+      }),
+      RCTDeviceEventEmitter.addListener('websocketErrored', ev => {
```

Comment:
What is the semantic difference between "errored" and "failed"? As a developer, when would I want to handle error and not failure or the other way round?

Looking at the implementation, this event is only raised from Java, so it's Android-only. iOS only has the failed event:
https://github.com/facebook/react-native/blob/master/Libraries/WebSocket/RCTWebSocketManager.m

Can you remove this event and always call "failed" if anything goes wrong? IMHO that API is easier to reason about.

Looking at the documentation of okhttp's on failure, the cases when this is raised don't seem to have a clear distinction from the places where you get an `IOException`: https://github.com/square/okhttp/blob/master/okhttp-ws/src/main/java/com/squareup/okhttp/ws/WebSocketListener.java#L42